File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-5.cpp
// /**
//  * Definition for a binary tree node.
//  * struct TreeNode {
//  *     int val;
//  *     TreeNode *left;
//  *     TreeNode *right;
//  *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//  *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//  *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
//  * };
//  */

// class Solution {
// public:
//     bool isEqual(TreeNode* tree1, TreeNode* tree2) {
//         if (tree1 == tree2) return true;
//         if (!(tree1 && tree2)) return false;

//         return (tree1->val == tree2->val) && isEqual(tree1->left, tree2->left) &&
//                 isEqual(tree1->right, tree2->right);

// }    
//     bool isSubtree(TreeNode* root, TreeNode* subRoot) {
//         if (!subRoot) return true;
//         if (!root) return false;

//         return isEqual(root, subRoot) ||
//             isSubtree(root->left, subRoot) ||
//             isSubtree(root->right, subRoot);
//     }
// };

class Solution {
public:
    string serialize(TreeNode* node) {
        if (!node) return "#,";

        return to_string(node->val) + "," +
               serialize(node->left) +
               serialize(node->right);
    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        string s1 = serialize(root);
        string s2 = serialize(subRoot);

        return s1.find(s2) != string::npos;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        ans = 0
        if root is None: return ans
        stack = [(root, root.val)]

        while stack:
            base, tracker = stack.pop()
            if base is None: continue
            if tracker <= base.val: ans += 1
            stack.append((base.left, max(tracker, base.val)))
            stack.append((base.right, max(tracker, base.val)))
        return ans
            